auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -671,7 +708,166 @@ private MethodSpec createReadMethod(ClassName className,
+        .addException(IOException.class);
+
+    ClassName token = ClassName.get(JsonToken.NULL.getDeclaringClass());
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("return null");
+    readMethod.endControlFlow();
+
+
+    // Instantiate the AutoValue'd class builder
+    FieldSpec builder = FieldSpec.builder(ClassName.get(builderContext.builderType()), "builder")
+        .build();
+
+    Set<ExecutableElement> builderMethods = builderContext.builderMethods();
+
+    if (builderMethods.size() == 0) {
+      // If no builder method defined, instantiate directly.
+      readMethod.addStatement("$T $N = new $T.$L()", builder.type, builder, className,
+          builderContext.builderType().getSimpleName());
+    } else {
+      ExecutableElement builderMethod;
+      if (builderMethods.size() == 1) {
+        // If there is only 1, use it.
+        builderMethod = builderMethods.stream().findFirst().get();
+      } else {
+        // Otherwise, find the only builder method that is annotated.
+        Set<ExecutableElement> annotatedMethods = builderMethods.stream()
+            .filter(e -> MoreElements.isAnnotationPresent(e, AutoValueGsonBuilder.class))
+            .collect(Collectors.toSet());
+
+        if (annotatedMethods.size() == 0) {
+          processingEnvironment.getMessager()
+              .printMessage(
+                  Diagnostic.Kind.ERROR,
+                  "Too many builder methods in " + className.simpleName() + ". Annotate the "
+                      + "builder this TypeAdapter should use with @AutoValueGsonBuilder."
+              );
+          throw new IllegalStateException();
+        } else if (annotatedMethods.size() == 1) {
+          builderMethod = annotatedMethods.stream().findFirst().get();
+        } else {
+          processingEnvironment.getMessager()
+              .printMessage(
+                  Diagnostic.Kind.ERROR,
+                  "Too many @AutoValueGsonBuilder annotations",
+                  annotatedMethods.stream().findFirst().get()
+              );
+          throw new IllegalStateException();
+        }
+      }
+
+      readMethod.addStatement("$T $N = $T.$L", builder.type, builder, autoValueClassName, builderMethod);
+    }
+
+    // begin reading object
+    readMethod.addStatement("$N.beginObject()", jsonReader);
+    readMethod.beginControlFlow("while ($N.hasNext())", jsonReader);
+
+    FieldSpec name = FieldSpec.builder(String.class, "_name").build();
+    readMethod.addStatement("$T $N = $N.nextName()", name.type, name, jsonReader);
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("continue");
+    readMethod.endControlFlow();
+
+    readMethod.beginControlFlow("switch ($N)", name);
+    for (Property prop : properties) {
```

Comment:
A lot of this is duplicated. Could we maybe share this and just have a notion of "assignment" codeblocks that either set the builder property or assign to the local variable if there's no builder?

Reply:
Ya, let me take a stab at that